compiler/lib/backend/optree/optimizer/transforms/control_flow_sink.cpp

```diff
@@ -0,0 +1,142 @@
+#include <algorithm>
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <string_view>
+#include <vector>
+
+#include "compiler/optree/adaptors.hpp"
+#include "compiler/optree/operation.hpp"
+#include "compiler/utils/helpers.hpp"
+
+#include "optimizer/opt_builder.hpp"
+#include "optimizer/transform.hpp"
+
+using namespace optree;
+using namespace optree::optimizer;
+
+namespace {
+
+struct ControlFlowSinkOps : public Transform<FunctionOp> {
+    using Transform::Transform;
+
+    std::string_view name() const override {
+        return "ControlFlowSinkOps";
+    }
+
+    struct Region {
+        uint32_t id;
+        Operation::Ptr owner;
+    };
+    using RegionMap = std::map<Operation::Ptr, Region>;
+
+    void fillRegionMap(const Operation::Ptr &op, RegionMap &regionMap, uint32_t &scope) const {
+        if (!op->body.empty())
+            scope++;
+        for (auto &child : op->body) {
+            regionMap[child] = {scope, op};
+            fillRegionMap(child, regionMap, scope);
+        }
+    };
+
+    bool isParentChild(const Operation::Ptr &child, const Operation::Ptr &parentCandidate) const {
+        if (child == parentCandidate)
+            return true;
+        auto parent = child->parent;
+        while (parent) {
+            if (parent == parentCandidate)
+                return true;
+            parent = parent->parent;
+        }
+        return false;
+    }
+
+    void sinkOperation(const Operation::Ptr &child, RegionMap &regionMap, OptBuilder &builder) const {
+        uint32_t childPos = regionMap[child].id;
+        if (child->results.size() != 1)
+            return;
+        for (const auto &result : child->results) {
+            std::vector<Region> usingIn{}; // can use set in this?
+            bool notFound = false;
+            for (const auto &use : result->uses) {
+                auto user = use.lock();
+                if (auto search = regionMap.find(user); search != regionMap.end()) {
+                    if (childPos < search->second.id)
+                        usingIn.emplace_back(search->second);
+                    else
+                        notFound = true;
+                } else {
+                    notFound = true;
+                }
+            }
+
+            if (!notFound && !usingIn.empty()) {
+                auto minRegion = usingIn[0];
+                const auto &candidate = usingIn[0].owner;
+                // cheking that operations are family
+                bool isBrothers = false;
```

Comment:
для этого есть более подходящее слово

```suggestion
                bool siblings = false;
```

Reply:
Братанов на siblings не меняют